text/00110-compatibility-strategy.md

```diff
@@ -0,0 +1,364 @@
+
+### Things to note
+
+- The expected error message in the case of incompatibility is generic. It should not be determined by the developer. This will ensure consistency and correctness.
+- When a version is deemed incompatible, this test needs to be amended with another `if` clause that puts that version in a different flow. We call this process, along with actually bumping the version, `mark-incompatible`. Remember this phrase, as you will see it again a bit further down.
+
+\+ **test-framework-compatibility.sh**
+
+```bash
+source ./common.bash
+
+function run_tests() {
+
+    expected_error=$1
+
+    run_integration_tests_perm ${PREVIOUS_VERSION} ${CURRENT_VERSION} ${PREVIOUS_VERSION} ${expected_error}
+}
+
+function expect_error() {
+    run_tests "A newer version of the CDK CLI (>= ${NEXT_VERSION}) is necessary to interact with this app"
+}
+
+function expect_pass() {
+    run_tests
+}
+
+if [ ${CURRENT_VERSION} == "1.19.0" ]; then
+  
+    # version 1.20.0 of the framework bumped protocol version, so we expect an error message.
+    expect_error
+
+elif [ ${CURRENT_VERSION} == "1.20.0" ]; then
+
+    # version 1.21.0 of the framework bumped protocol version, so we expect an error message.
+    expect_error
+
+else
+
+    # common case, backwards compatible (i.e no protocol version bump)
+    # all old integration tests should pass.
+    expect_pass
+
+fi
+
+```
+
+In addition, we create a new file, which is responsible for validating the compatibility of new cli versions against old framework versions.
+
+### Things to note
+
+- Nothing really :)
+
+\+ **test-cli-compatibility.sh**
+
+```bash
+source ./common.bash
+
+# CLI should always support older framework versions, so these tests should always pass.
+run_integration_tests_perm ${PREVIOUS_VERSION} ${PREVIOUS_VERSION} ${CURRENT_VERSION}
+
+```
+
+## Developer Workflow
+
+Ideally, developers should mainly focus on code changes that are relevant to the functionality they are implementing, not necessarily concerning themselves with versioning issues. 
+
+This design attempts to provide a mechanism that, given the compatibility is broken, will guide the developer to the necessary code changes that need to be done. This way, developers don't have to *remember* anything, they simply won't be able to properly merge their changes until those versioning concerns are addressed.
+
+> Thats not to say we should stop thinking about compatibility when designing our changes, it simply alleviates the fear of unintentionally breaking compatibility.
+
+Following are the two possible scenarios that might break compatibility, with a diagram that represents the actions the developer would take in order for the change to be accepted. 
+
+### Introducing changes to the CLI
+
+This is really straight forward, new cli versions must always be compatible with older framework versions.
+
+1. Developer merges a PR to master.
+2. CLI compatibility tests are executed.
```

Comment:
We need to give contributors tools to validate compatibility before they merge

Reply:
This would be same as validating regular integration tests, which we now instruct to run locally.